library/src/lapack/roclapack_syevx_heevx.hpp

```diff
@@ -42,42 +42,40 @@ ROCSOLVER_KERNEL void __launch_bounds__(BS1) syevx_sort_eigs(const rocblas_int n
     T* Z = load_ptr_batch<T>(ZZ, bid, shiftZ, strideZ);
     rocblas_int* ifail = ifailA + (bid * strideIfail);
 
-    if(tid < n)
+    for(j = 0; j < nev - 1; j++)
     {
-        for(j = 0; j < nev - 1; j++)
+        i = 0;
+        S tmp1 = W[j];
+        for(jj = j + 1; jj < nev; jj++)
         {
-            i = 0;
-            S tmp1 = W[j];
-            for(jj = j + 1; j < nev; j++)
+            if(W[jj] < tmp1)
             {
-                if(W[jj] < tmp1)
-                {
-                    i = jj;
-                    tmp1 = W[jj];
-                }
+                i = jj;
+                tmp1 = W[jj];
             }
-            __syncthreads();
+        }
+        __syncthreads();
+
+        if(i != 0)
+        {
+            if(tid == 0)
+            {
+                W[i] = W[j];
+                W[j] = tmp1;
+            }
+            
+            for(int k = tid; k < n; k += hipBlockDim_x)
+                swap(Z[k + i * ldz], Z[k + j * ldz]);
 
-            if(i != 0)
+            if(tid < info)
             {
-                if(tid == 0)
-                {
-                    W[i] = W[j];
-                    W[j] = tmp1;
-                }
-
-                swap(Z[tid + i * ldz], Z[tid + j * ldz]);
-
-                if(tid < info)
-                {
-                    if(ifail[tid] == i + 1)
-                        ifail[tid] = j + 1;
-                    else if(ifail[tid] == j + 1)
-                        ifail[tid] = i + 1;
-                }
+                if(ifail[tid] == i + 1)
+                    ifail[tid] = j + 1;
+                else if(ifail[tid] == j + 1)
+                    ifail[tid] = i + 1;
             }
```

Comment:
I suspect it will be very rare, but since `tid < BS1` now, it looks like there will be some ifail entries that are missed when `info >= BS1`. It might be a good idea to change the if into a loop, i.e.
for(k = tid; k < info; k += hipBlockDim_x)

Reply:
Right! Better to change that...